### ISSUE TITLE
Support job-(pre|post)-start commands

### ISSUE THREAD

**Issue by CAFxX** [cloudfoundry/bosh#1688]
We would find very helpful to have the ability to specify job (pre|post)-start commands.
Consider a job A with 2 instances: right now we have the following events:

- *other jobs*
- unmonitor A/1
- drain A/1
- stop A/1
- pre-start A/1
- monit start A/1
- post-start A/1
- unmonitor A/2
- drain A/2
- stop A/2
- pre-start A/2
- monit start A/2
- post-start A/2
- *other jobs*
- post-deploy (all jobs)

what we would like to have is a job-(pre|post)-start event that run respecitively immediately before the first instance of a job is updated and immediately after the last instance of a job is updated:

- *other jobs*
- **job-pre-start A**
- unmonitor A/1
- drain A/1
- stop A/1
- pre-start A/1
- monit start A/1
- post-start A/1
- unmonitor A/2
- drain A/2
- stop A/2
- pre-start A/2
- monit start A/2
- post-start A/2
- **job-post-start A**
- *other jobs*
- post-deploy (all jobs)

We don't have preferences about which job instance should run the job-(pre|post)-start script.

# Use-cases
The two main use cases for this feature are:
- for "cluster-like" jobs, ability to change the runtime cluster configuration immediately before starting the update and immediately after finishing the update (e.g. elasticsearch recommends changing runtime configuration to minimize disruption during rolling updates)
- for any job, ability to inform monitoring system(s) that a component is being serviced (to avoid false positives)

**Comment by apoydence** [comment]
@CAFxX is this not covered by [pre-start](https://bosh.io/docs/pre-start.html) and [post-start](https://bosh.io/docs/post-start.html) scripts?

**Comment by CAFxX** [comment]
@apoydence I included pre-start and post-start in the list above, as you can see they don't match the proposed job-pre-start and job-post-start. Maybe a better name would be instance-group-pre-start and instance-group-post-start?

**Comment by apoydence** [comment]
@CAFxX Ah, I see what you are proposing now. Thank you for the clarification.

**Comment by jfmyers9** [comment]
Hi @CAFxX,

Thanks for the issue. At the moment we aren't planning to implement a "pre/post-instance" lifecycle hook. In the future, we are discussing the ability to allow operators to have more fine grained control over when errands and services execute during a deploy. We hope that this future feature will provide the ability to have a more complex deploy process.

As a result, I am going to close this issue. Feel free to reopen / create a new issue if you have further questions.

Best,

@jfmyers9, CF BOSH

**Comment by CAFxX** [comment]
> We hope that this future feature will provide the ability to have a more complex deploy process.

Is there a proposal I can comment on, to make sure this doesn't get lost? @jfmyers9

**Comment by dpb587-pivotal** [comment]
I think this closest thing we have for you right now is a prior note about this sort of thing in [bosh-notes](https://github.com/cloudfoundry/bosh-notes/blob/master/proposals/errand-triggers.md). We'll probably be communicating a bit more about it if we get closer to a recommendation for change.
